building_sim_plugins/building_gazebo_plugins/src/crowd_simulator.cpp

```diff
@@ -0,0 +1,273 @@
+//WorldPlugin
+void CrowdSimulatorPlugin::Load(gazebo::physics::WorldPtr world,
+  sdf::ElementPtr sdf)
+{
+  _world = world;
+  
+  if (!_crowdSimInterface->readSDF(sdf)) {
+    exit(EXIT_FAILURE);
+  }
+
+  if (!_crowdSimInterface->initCrowdSim()) {
+    RCLCPP_ERROR(_crowdSimInterface->logger(), "Crowd simulation failed to initialize.");
+    exit(EXIT_FAILURE);
+  }
+
+  if (!_spawnAgentsInWorld()) {
+    RCLCPP_ERROR(_crowdSimInterface->logger(), "Crowd simulation failed to spawn agents in the world.");
+    exit(EXIT_FAILURE);
+  }
+
+  _updateConnectionPtr = gazebo::event::Events::ConnectWorldUpdateBegin(
+    [this](gazebo::common::UpdateInfo updateInfo)
+    {
+      _Update(updateInfo);
+    }
+  );
+}
+
+//============================================
+void CrowdSimulatorPlugin::_Update(const gazebo::common::UpdateInfo& updateInfo)
+{
+  //first round do nothing, initialize time stamp
+  if (_lastSimTime == gazebo::common::Time::Zero)
+  {
+    _lastSimTime = updateInfo.simTime;
+    _lastTime = updateInfo.simTime;
+  }
+
+  if (!_initialized)
+  {
+    // not initizalied
+    _initSpawnedAgents();
+    return;
+  }
+
+  auto deltaTime = (updateInfo.simTime - _lastTime).Double();
+  _lastTime = updateInfo.simTime;
+
+  auto deltaSimTime = (updateInfo.simTime - _lastSimTime).Double();
+  if (deltaSimTime - _crowdSimInterface->getSimTimeStep() < 1e-6)
+  {
+    deltaSimTime = 0.0;
+  }
+  else
+  {
+    _lastSimTime = updateInfo.simTime;
+    _crowdSimInterface->oneStepSim();
+  }
+  _UpdateAllObjects(deltaTime, deltaSimTime);
+
+}
+
+//============================================
+void CrowdSimulatorPlugin::_UpdateAllObjects(double deltaTime, double deltaSimTime) 
+{
+  for (size_t id = 0; id < _objectsCount; id++)
+  {
+    ObjectPtr objPtr = _crowdSimInterface->getObjectById(id);
+    gazebo::physics::ModelPtr modelPtr = _world->ModelByName(objPtr->modelName);
+
+    //update external agents
+    if (objPtr->agentPtr->_external)
+    {
+      ignition::math::Pose3d pose = modelPtr->WorldPose();
+      _crowdSimInterface->updateExternalAgent<ignition::math::Pose3d>(objPtr->agentPtr, pose);
+      continue;
+    }
+
+    //update internal agents
+    //not yet reach the simulation time step, the internal agent position only updated at simulation time step
+    if(deltaSimTime - 0.0 < 1e-6) continue;
+    auto typePtr = _crowdSimInterface->_modelTypeDBPtr->Get(objPtr->typeName);
+    _UpdateInternalObject(deltaTime, deltaSimTime, objPtr->agentPtr, modelPtr, typePtr);
+  }
+}
+
+//============================================
+void CrowdSimulatorPlugin::_UpdateInternalObject(double deltaTime, double deltaSimTime,
+  const crowd_simulator::AgentPtr agentPtr,
+  const gazebo::physics::ModelPtr modelPtr,
+  const crowd_simulator::ModelTypeDatabase::RecordPtr typePtr)
+{
+
+  if (!agentPtr)
+  {
+    RCLCPP_ERROR(_crowdSimInterface->logger(), "Null agentPtr when update Object!");
+    assert(agentPtr);
```

Comment:
Is there a reason the `assert` is in this check? Also do you intend to have the code continue executing even if the pointer is invalid? My suggestion would be to check for `nullptr` and `return`,
```
if (!agentPtr)
{
  RCLCPP_ERROR(_crowdSimInterface->logger(), "Null agentPtr when update Object!");
  return;
}
```

Or if this check is really for debugging purposes, maybe you can get away with only using a single `assert` call instead of an `if` case,
```
assert(agentPtr);
assert(modelPtr);
```

Reply:
Thanks for pointing out. Yep I remember I wrote this assert in the early stage to abort the plugin once an invalid pointer is found, which I think `assert()` is doing the check, and if found false, assert will call `abort()`. Just return from this function might just get a log and the rest code is continuing. 
Is there any way to do it elegantly? maybe I should use `exit()` ?

Reply:
There is not that clear a solution I believe. It will ultimately depends on how you would like the plugin's behaviour to be, and how often would you expect these edge cases to show up. If it is something that is critical for the continual operation of the plugin, you might want to crash the plugin while providing error logs. Otherwise if this happens once in a while, as a result of dropping a sensor message for example, and you would like the plugin to continue, then an error log while calling `return;` on the function would be good.